libtock/alarm_timer.c

```diff
@@ -4,11 +4,131 @@
+
+  // This calculation is mathematically equivalent to doing:
+  //
+  //   `leftover_ticks` / (`frequency` / `milliseconds_per_second`)
+  //
+  // This is taking the remaining unconverted ticks (leftover_ticks)
+  // and dividing by the number of ticks per millisecond
+  // (`frequency` (ticks per second) / `1000` milliseconds per second)
+  // The division is done this way because of the same argument in
+  // `ms_to_ticks`.
+  milliseconds += (leftover_ticks * milliseconds_per_second) / frequency;
+  return milliseconds;
+}
+
+// WARNING: This function can silently overflow if the output
+// number of ticks is larger than `UINT32_MAX`.
+static uint32_t ms_to_ticks(uint32_t ms) {
+  // This conversion is accurate to within 1 millisecond of a true
+  // fractional conversion.
+  uint32_t frequency;
+  alarm_internal_frequency(&frequency);
+
+  // Note the following code is mathematically equivalent to the following
+  // more readable conversion.
+  // ```
+  //   uint32_t seconds         = ms / 1000;
+  //   uint32_t leftover_millis = ms % 1000;
+  //   uint32_t millihertz      = frequency / 1000; // ticks per millisecond
+  //   return (seconds * frequency) + (leftover_millis * millihertz);
+  // ```
+  // We can check the logical correctness of this conversion by doing dimensional
+  // analysis of the units.
+  // For the first half, `seconds` (sec) * `frequency` (ticks per second) = a quantity in ticks.
+  // And for the second half, `leftover_millis` (milliseconds) * `millihertz` (ticks per millisecond)
+  // = a quantity in ticks.
+  // Thus, we know the calculation is logically correct as the units cancel out to return ticks.
+  //
+  // Although these are logically equivalent, this human readable implementation is
+  // faulty. I will prove this by (1.) showing an upper bound on the error of this conversion,
+  // and then (2.) explaining why this commented human readable implementation is
+  // equivalent to the true implementation.
+  //
+  // Part 1:
+  // The splitting of the input milliseconds has no error.
+  uint32_t seconds         = ms / 1000;
+  uint32_t leftover_millis = ms % 1000;
+  // In other words, `seconds` + `leftover_millis` is always == `ms`.
+
+  // The first half of the output converts the seconds component to ticks with no error.
+  // This multiplication has no error because the board frequency is given in Hertz (ticks per second).
+  uint32_t ticks = seconds * frequency;
+
+  // Now that we have converted the full seconds part of `ms` to ticks, we must
+  // convert the `leftover_millis`.
+  // The reason for splitting this conversion rather than doing a direct conversion
+  // is due to 32-bit overflow. `leftover_millis` can only take values in the range
+  // [0, frequency). Thus, this multiplication is upper bounded by `frequency`^2. This
+  // calculation will not overflow as long as
+  //   `frequency`^2 < `UINT32_MAX` -> frequency < 65535.99 -> frequency <= 2^16 - 1
```

Comment:
Same as above, I think we want this to support frequencies at least up to 100MHz considering the kernel currently provides enum variants for frequencies up to 100MHz.

I do not think there is a way around using a `u64` here.

Reply:
This is fixed in the separated PR. Also, I made an error in the calculation here which is also fixed.